File: plugin/wrapper/scripts/groklib/deny_write.py
# wrapper/scripts/groklib/deny_write.py
#
# Protected write-deny globs + match SSOT for direct finalize and (via Node
# mirror) auto/peer apply pre-block. Data lives in
# plugin/references/deny-write-globs.json - never hardcode a second list here.

from __future__ import annotations

import fnmatch
import json
import pathlib
from typing import List, Sequence, Tuple

# plugin/references/deny-write-globs.json relative to this module:
# groklib/ -> scripts/ -> wrapper/ -> plugin/ -> references/
_DATA_PATH = (
    pathlib.Path(__file__).resolve().parents[3]
    / "references"
    / "deny-write-globs.json"
)

_LOADED = False
_GLOBS: Tuple[str, ...] = ()


def _load_ssot() -> Tuple[str, ...]:
    global _LOADED, _GLOBS
    if _LOADED:
        return _GLOBS
    if not _DATA_PATH.is_file():
        raise RuntimeError(
            "deny-write SSOT missing at {}".format(_DATA_PATH)
        )
    with _DATA_PATH.open("r", encoding="utf-8") as fh:
        doc = json.load(fh)
    globs = doc.get("globs")
    if not isinstance(globs, list) or not globs:
        raise RuntimeError("deny-write SSOT has empty/invalid globs")
    if not all(isinstance(g, str) and g for g in globs):
        raise RuntimeError("deny-write SSOT globs must be non-empty strings")
    _GLOBS = tuple(globs)
    _LOADED = True
    return _GLOBS


def deny_write_ssot_path() -> pathlib.Path:
    """Absolute path to the shared deny-write JSON SSOT (for tests/docs)."""
    return _DATA_PATH


def deny_write_globs() -> Tuple[str, ...]:
    """Return the deny-write glob tuple from the JSON SSOT (cached)."""
    return _load_ssot()


# Compat alias used by direct_finalize and historical imports.
# Populated eagerly so ``from groklib.deny_write import DENY_WRITE_GLOBS`` works
# and remains a tuple identical to the previous in-module constant.
DENY_WRITE_GLOBS: Tuple[str, ...] = deny_write_globs()


def posix_rel(path: str) -> str:
    """POSIX-normalize a repo-relative path WITHOUT stripping a leading dotfile.

    Do not use ``str.lstrip('./')``: that treats '.' as a character class and
    would turn ``.env`` into ``env`` and ``.git/config`` into ``git/config``.
    """
    norm = str(path or "").replace("\\", "/")
    while norm.startswith("./"):
        norm = norm[2:]
    return norm


def path_matches_deny(path: str, globs: Sequence[str] | None = None) -> bool:
    """True when a repo-relative path matches the deny-write globs or any .git.

    Algorithm (shared with Node deny-write.mjs / golden vectors):
      1. POSIX-normalize (strip leading ``./`` only).
      2. Empty path -> False.
      3. Any path component named ``.git`` -> True (root, nested vendor/lib/.git,
         modules).
      4. Else match each glob via fnmatch against full path OR basename
         (Python fnmatch: ``*`` matches ``/``).
    """
    patterns: Sequence[str] = _load_ssot() if globs is None else globs
    norm = posix_rel(path)
    if not norm:
        return False
    parts = [p for p in norm.split("/") if p]
    # Any path component named .git (root, nested vendor repo, submodule gitdir).
    if ".git" in parts:
        return True
    base = parts[-1] if parts else ""
    for pattern in patterns:
        if fnmatch.fnmatch(norm, pattern) or fnmatch.fnmatch(base, pattern):
            return True
    return False


def load_match_vectors() -> List[dict]:
    """Return matchVectors from the SSOT (for dual-host golden tests)."""
    with _DATA_PATH.open("r", encoding="utf-8") as fh:
        doc = json.load(fh)
    vectors = doc.get("matchVectors") or []
    if not isinstance(vectors, list):
        raise RuntimeError("deny-write SSOT matchVectors must be a list")
    return vectors

File: plugin/scripts/lib/deny-write.mjs
// plugin/scripts/lib/deny-write.mjs
//
// Protected write-deny globs + match SSOT for the shared auto/peer apply
// pre-block. Data: plugin/references/deny-write-globs.json (parity with Python
// groklib.deny_write / direct_finalize.path_matches_deny). No second list.

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
export const DENY_WRITE_SSOT_PATH = path.resolve(
  HERE,
  "../../references/deny-write-globs.json"
);

let _globs = null;
let _doc = null;

/**
 * Load and cache the deny-write SSOT JSON. Fail closed if missing/invalid.
 * @returns {{schemaVersion: number, globs: string[], matchVectors?: object[]}}
 */
export function loadDenyWriteSsot() {
  if (_doc) return _doc;
  if (!fs.existsSync(DENY_WRITE_SSOT_PATH)) {
    throw new Error(`deny-write SSOT missing at ${DENY_WRITE_SSOT_PATH}`);
  }
  const raw = fs.readFileSync(DENY_WRITE_SSOT_PATH, "utf8");
  const doc = JSON.parse(raw);
  if (!Array.isArray(doc.globs) || doc.globs.length === 0) {
    throw new Error("deny-write SSOT has empty/invalid globs");
  }
  if (!doc.globs.every((g) => typeof g === "string" && g)) {
    throw new Error("deny-write SSOT globs must be non-empty strings");
  }
  _doc = doc;
  _globs = doc.globs.slice();
  return _doc;
}

/**
 * @returns {string[]}
 */
export function denyWriteGlobs() {
  loadDenyWriteSsot();
  return _globs.slice();
}

/** Compat export (tuple-like array) for tests mirroring Python DENY_WRITE_GLOBS. */
export function getDenyWriteGlobs() {
  return denyWriteGlobs();
}

/**
 * POSIX-normalize a repo-relative path WITHOUT stripping a leading dotfile.
 * Mirrors Python groklib.deny_write.posix_rel.
 * @param {string} p
 * @returns {string}
 */
export function posixRel(p) {
  let norm = String(p || "").replace(/\\/g, "/");
  while (norm.startsWith("./")) norm = norm.slice(2);
  return norm;
}

/**
 * Python-fnmatch-compatible matcher for the patterns we ship (literals + *).
 * ``*`` matches any string including ``/`` (Python stdlib fnmatch parity).
 * @param {string} name
 * @param {string} pattern
 * @returns {boolean}
 */
export function fnmatchStar(name, pattern) {
  let out = "^";
  for (let i = 0; i < pattern.length; i++) {
    const c = pattern[i];
    if (c === "*") {
      out += ".*";
      continue;
    }
    if (c === "?") {
      out += ".";
      continue;
    }
    if ("\\.^$+{}()|[]".includes(c)) out += `\\${c}`;
    else out += c;
  }
  out += "$";
  return new RegExp(out, "s").test(name);
}

/**
 * True when a repo-relative path matches the deny-write globs or any path
 * component is `.git` (root, nested vendor/lib/.git, modules). Same algorithm
 * as Python path_matches_deny / golden vectors.
 * @param {string} pathStr
 * @param {string[]} [globs]
 * @returns {boolean}
 */
export function pathMatchesDeny(pathStr, globs) {
  const patterns = globs == null ? denyWriteGlobs() : globs;
  const norm = posixRel(pathStr);
  if (!norm) return false;
  const parts = norm.split("/").filter(Boolean);
  // Any component named .git (root, nested vendor repo, submodule gitdir).
  if (parts.includes(".git")) return true;
  const base = parts[parts.length - 1] || "";
  for (const pattern of patterns) {
    if (fnmatchStar(norm, pattern) || fnmatchStar(base, pattern)) return true;
  }
  return false;
}

/**
 * First deny-matched path from a touch set (sorted for stable reporting), or null.
 * @param {Iterable<string>} paths
 * @returns {string[]}
 */
export function protectedPathsIn(paths) {
  const offenders = [];
  for (const p of paths) {
    if (pathMatchesDeny(p)) offenders.push(p);
  }
  return offenders.sort();
}
